Sources/AsyncHTTPClient/Utils.swift

```
@@ -77,149 +78,207 @@ extension NIOClientTCPBootstrap {
-                }
+    if requiresTLS, configuration.proxy != nil {
+        let handshakePromise = channel.eventLoop.makePromise(of: Void.self)
+        channel.pipeline.syncAddLateSSLHandlerIfNeeded(for: key,
+                                                       sslContext: sslContext!,
+                                                       handshakePromise: handshakePromise)
+        handshakeFuture = handshakePromise.futureResult
+    } else if requiresTLS {
+        do {
+            handshakeFuture = try channel.pipeline.syncOperations.handler(type: TLSEventsHandler.self).completionPromise.futureResult
+        } catch {
+            return channel.eventLoop.makeFailedFuture(error)
+        }
+    } else {
+        handshakeFuture = channel.eventLoop.makeSucceededVoidFuture()
+    }
 
-                try syncOperations.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
+    return handshakeFuture.flatMapThrowing {
+        let syncOperations = channel.pipeline.syncOperations
 
-                #if canImport(Network)
-                    if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *), bootstrap.underlyingBootstrap is NIOTSConnectionBootstrap {
-                        try syncOperations.addHandler(HTTPClient.NWErrorHandler(), position: .first)
-                    }
-                #endif
-
-                switch configuration.decompression {
-                case .disabled:
-                    ()
-                case .enabled(let limit):
-                    let decompressHandler = NIOHTTPResponseDecompressor(limit: limit)
-                    try syncOperations.addHandler(decompressHandler)
-                }
+        // If we got here and we had a TLSEventsHandler in the pipeline, we can remove it ow.
+        if requiresTLS {
+            channel.pipeline.removeHandler(name: TLSEventsHandler.handlerName, promise: nil)
+        }
```

I think we should make this sync as well. If we want to ignore the error go with: `try?`, but I doubt that we would like to ignore the error.

---

You can't synchronously remove a handler. The handler is on control of removals (through the `RemovableChannelHandler` protocol and they may be async.